src/08-recommendations/03-pipeline-recommend.ts

```diff
@@ -0,0 +1,325 @@
+    temperature: 0,
+    messages,
+    tools,
+  })
+  VERBOSE && console.log(chalk.green("\n< Output <\n"))
+  VERBOSE && console.log(chalk.green.dim(JSON.stringify(response, null, 2)))
+
+  return response
+}
+
+function isTool(response: OpenAI.Chat.Completions.ChatCompletion) {
+  return response.choices[0].finish_reason === "tool_calls"
+}
+
+async function getToolResult(
+  response: OpenAI.Chat.Completions.ChatCompletion
+): Promise<OpenAI.Chat.ChatCompletionToolMessageParam> {
+  console.log(chalk.black("\nGetting tool result from Artsy"))
+
+  const { tool_calls: toolCalls } = response.choices[0].message
+
+  VERBOSE && console.log(chalk.cyan.dim("\n> Input >\n"))
+  VERBOSE && console.log(chalk.cyan.dim(JSON.stringify(toolCalls, null, 2)))
+
+  const firstToolCall = toolCalls![0]
+
+  const { id } = firstToolCall
+  const { name } = firstToolCall.function
+
+  let result: unknown
+
+  if (name === "getArtworkRecommendations") {
+    result = await getArtworkRecommendations()
+  } else {
+    throw new Error(`Unknown tool call: ${name}`)
+  }
+
+  VERBOSE && console.log(chalk.magenta("\n< Tool result <\n"))
+  VERBOSE && console.log(chalk.magenta.dim(JSON.stringify(result, null, 2)))
+
+  return {
+    role: "tool",
+    tool_call_id: id,
+    content: JSON.stringify(result, null, 2),
+  }
+}
+
+async function getArtworkRecommendations() {
+  const query = `
+    query ArtworkRecommendations($userID: String!, $size: Int!, $includeBackfill: Boolean!) {
+      who: me {
+        email
+      }
+      artworksForUser(userId: $userID, includeBackfill: $includeBackfill, first: $size) {
+        edges {
+          node {
+            ...artwork
+          }
+        }
+      }
+      me {
+        artworkRecommendations(first: $size) {
+          edges {
+            node {
+              ...artwork
+            }
+          }
+        }
+        recommendedArtworks(first: $size) {
+          edges {
+            node {
+              ...artwork
+            }
+          }
+        }
+      }
+    }
+
+    fragment artwork on Artwork {
+      slug
+      title
+      date
+      artist {
+        name
+        gender
+        nationality
+        birthday
+      }
+      rarity: attributionClass {
+        name
+      }
+      medium: mediumType {
+        name
+      }
+      materials: medium
+      price: saleMessage
+      # dominantColors
+      # href
+    }
+  `
+
+  const variables = {
+    userID: userID,
+    size: NUM_CANDIDATES_PER_FIELD,
+    includeBackfill: INCLUDE_BACKFILL,
+  }
+
+  const headers = {
+    "X-Access-Token": accessToken!,
+  }
+
+  const response = await metaphysics({ query, variables, headers })
+
+  // hard to know which recommendation fields will be populated,
+  // so for now just fetch them all and concatenate them
+  const artworksToUse = [
+    ...response.data.me.recommendedArtworks.edges,
+    ...response.data.me.artworkRecommendations.edges,
+    ...response.data.artworksForUser.edges,
+  ]
```

Comment:
I asked about this [here](https://artsy.slack.com/archives/C033E79AQG4/p1715869812362219), still a little murky on the details.